docs/design/02-topologies.md

```diff
@@ -0,0 +1,288 @@
+  the repo;
+
+- having these built in to Metalware means we won't need to duplicate
+  validations between repos or places within repos;
+
+- we may later want to do non-regex validation, e.g. a `dns` type could
+  validate that a particular IP entered gives the expected DNS response (though
+  this particular example sounds like overkill).
+
+
+# `metal configure` commands
+
+The `metal configure` commands each take their specified questions and ask
+them, save the results to a particular file within a new
+`/var/lib/metalware/config/` directory, and perform any other actions
+appropriate to the resource being configured.
+
+## `metal configure all`
+
+Asks the `all` questions; the validated, converted answers will be saved to
+`/var/lib/metalware/config/all.yaml` in the following format:
+
+```yaml
+answers:
+  example_question_identifier: 7
+  another_example: 'foo'
+```
+
+## `metal configure group $GROUP`
+
+Asks for a genders line to be entered first, then similarly to `all` asks the
+`group` questions; the results will then be saved to
+`/var/lib/metalware/config/groups/$GROUP.yaml`
+
+```yaml
+genders: 'somenodes[01-03] somenodes,cluster,all'
+answers:
+  another_example: 'foo'
+```
+
+Note that initially the genders line will just take a line of text, later we
+can add more validation and potentially generate this based on e.g. entering a
+number of nodes in the group and a group name.
+
+Every time `metal configure group` is run we should regenerate
+`/opt/metalware/etc/genders` by (for now at least) simply concatenating all the
+group gender lines together; this way it should always be in sync with the
+configured groups. Aside: it may be worth including a line in the generated
+genders indicating this and that any changes to it directly may not persist.
+
+## `metal configure node $NODE`
+
+Similarly asks the questions configured for `node`, and saves these to
+`/var/lib/metalware/config/nodes/$NODE.yaml`. Note that unlike `configure all`,
+required before any nodes/group can be built, or `configure group`, required
+before a particular group can be built, it will not be required to `configure
+node` before building a particular node; `configure node` just provides a way
+to further specialize a particular node's configuration.
+
+## Re-running `configure` commands
+
+If any `configure` command is re-run we will also want to pre-fill the answer
+to each question with the saved answer.
+
+When saving the answers after re-running `configure`, we will want to make sure
+just the new answers are saved. For instance:
+
+- `metal configure all` run, an answer `'foo'` is given for the `some_question`
+  question;
+
+- the question `some_question` for `all` is removed from
+  `/var/lib/metalware/repo/configure.yaml`;
+
+- `metal configure all` is run again; `some_question` won't be asked and we
+  should make sure the previous answer is not still included in the resulting
+  config file.
```

Comment:
We should included a `reset` command that deletes `/var/lib/metalware/repo/configure.yaml`. Also we shouldn't alter the main `configure.yaml` file with the answers. We should save the answers in a separate file entirely. That way if we need to reset, we can delete the answer file and everything will be like it is brand new. 

We can still load both files to ensure we don't repeat questions that have all ready been asked.